### PR TITLE
SCHED-406: Fixing data structures and getting code ready for initial release

### PR DESCRIPTION
**Pull request by sraaphorst**
Trying to fix data structures so that we no longer use:
```
List[T]
```
for night data, but:
```
Dict[NightIndex, T]
```
In addition, many more cleanups, type hint adding, and modification of the code to use `NightIndices` so that the code is ready to simply iterate over nights to produce plans.

### CODE REVIEW

File: scheduler/core/components/optimizer/greedymax.py
Comment (stroncod) [inline scheduler/core/components/optimizer/greedymax.py:523]: I moved those to Plan (not being used yet) because I imagine with NIR science this should always happen so it should be done at the moment of the insertion

**Review by stroncod — APPROVED** [review]

File: scheduler/core/components/optimizer/__init__.py
Comment (sraaphorst) [reply scheduler/core/components/optimizer/__init__.py:39]: Right now, I'm not convinced that Bryan's code is iterating over nights except for using visibility calculations.
In the original code, he sets:
```
optimizer.periods = 3
```
but then when we call `optimizer.schedule`, it resets the `periods` to the length of the `time_grid`.

The initial code returns a set of three plans, but the second and third are always empty, so while GM is using `periods` nights for visibility calculations, it isn't (necessarily) actually calculating plans for `periods` nights.

Instead of doing that, what I do now is specify the `NightIndex` values for which we want to get a `Selection`. Since I haven't really changed any fundamental logic in `GreedyMax`, it is only using this logic for visibility calculations and calculating the plan for the first `NightIndex` in the `Selection`.

For  now, we need to focus on getting this out, so the way I see it, in `run_greedymax.py`, we iterate over the `NightIndex` subsets. Ultimately, what we will do is this, but using a loop over the `Selector.num_nights_to_schedule`:

```
plans = {}

selection = selector.select(night_indices=np.array([0, 1, 2]))
plans[0] = optimizer.schedule(selection)[0]
time accounting

selection = selector.select(night_indices=np.array([1, 2]))
plans[0] = optimizer.schedule(selection)[1]
time accounting

selection = selector.select(night_indices=np.array([2]))
plans[0] = optimizer.schedule(selection)[2]
time accounting
```

and then we will have plans for all nights across both sites.

To get GreedyMax to schedule plans for multiple nights in one call to `optimize` isn't the highest priority right now, and while I haven't had time to go over the logic in detail, since everything is scheduled on the first night and nothing is scheduled on nights 2 and 3, either in the original code or in the modifications, I think this is something we will have to look into later.

File: scheduler/core/components/optimizer/dummy.py
Comment (sraaphorst) [reply scheduler/core/components/optimizer/dummy.py:65]: No problem. 😄  I just went over all the code and tried to fix as many warnings as possible.
It was an easy fix... I figured it would take less time than creating a Jira task and iterating on it.

File: scheduler/core/components/optimizer/greedymax.py
Comment (sraaphorst) [reply scheduler/core/components/optimizer/greedymax.py:34]: Pretty sure this will ultimately have to change since Bryan said it is an issue in the `Atom` code with things not being set properly, but for now, this should do.

File: scheduler/core/components/optimizer/greedymax.py
Comment (sraaphorst) [reply scheduler/core/components/optimizer/greedymax.py:108]: I remember you saying you moved a bunch of these methods to lucupy... so after we merge this, if you want to remove them, let's go ahead and do that to clean up. For now I was just trying to get all the types in there when I could figure them out easily.

File: scheduler/core/components/optimizer/greedymax.py
Comment (sraaphorst) [reply scheduler/core/components/optimizer/greedymax.py:497]: Sounds good. Let's make a Jira task so that we come back to it in the next round of cleanups / fixes. 👍

File: scheduler/core/components/optimizer/greedymax.py
Comment (sraaphorst) [reply scheduler/core/components/optimizer/greedymax.py:523]: 👍

**Pull request merged**
